samza-core/src/main/java/org/apache/samza/operators/StreamGraphImpl.java

```diff
@@ -65,46 +71,51 @@ public StreamGraphImpl(ApplicationRunner runner, Config config) {
   }
 
   @Override
-  public <K, V, M> MessageStream<M> getInputStream(String streamId,
-      BiFunction<? super K, ? super V, ? extends M> msgBuilder) {
-    if (msgBuilder == null) {
-      throw new IllegalArgumentException("msgBuilder can't be null for an input stream");
-    }
+  public void setDefaultSerde(Serde<?> serde) {
+    Preconditions.checkNotNull(serde, "Default serde must not be null");
+    Preconditions.checkState(inputOperators.isEmpty() && outputStreams.isEmpty(),
+        "Default serde must be set before creating any input or output streams.");
+    this.defaultSerde = serde;
+  }
 
-    if (inputOperators.containsKey(runner.getStreamSpec(streamId))) {
-      throw new IllegalStateException("getInputStream() invoked multiple times "
-          + "with the same streamId: " + streamId);
-    }
+  @Override
+  public <M> MessageStream<M> getInputStream(String streamId, Serde<M> serde) {
+    Preconditions.checkNotNull(serde, "serde must not be null for an input stream.");
+    Preconditions.checkState(!inputOperators.containsKey(runner.getStreamSpec(streamId)),
+        "getInputStream must not be called multiple times with the same streamId: " + streamId);
 
     StreamSpec streamSpec = runner.getStreamSpec(streamId);
+    KV<Serde, Serde> kvSerdes = getKVSerdes(streamId, serde);
+    boolean isKeyedInput = serde instanceof KVSerde;
     inputOperators.put(streamSpec,
-        new InputOperatorSpec<>(streamSpec, (BiFunction<K, V, M>) msgBuilder, this.getNextOpId()));
+        new InputOperatorSpec<>(streamSpec, kvSerdes.getKey(), kvSerdes.getValue(), isKeyedInput, this.getNextOpId()));
```

Comment:
Can this `instance of` check be done in the constructor. You would n't need the extra param to the constructor then?

Same for `OutputStream`s

Reply:
These constructor doesn't get the KVSerde, only the key and value serdes separately. This is because the defaulting for the key/message serde based on the default serde happens here. Maybe there's another way to avoid this?

Reply:
makes sense, probably not worth doing then